zstd.cc

```diff
@@ -75,34 +99,47 @@ zstd_processor::zstd_processor(const opt_getter& opts)
     // We assume that the uncompressed input length is always <= chunk_len.
     auto cparams = ZSTD_getCParams(_compression_level, chunk_len, 0);
     auto cctx_size = ZSTD_estimateCCtxSize_usingCParams(cparams);
-    // According to the ZSTD documentation, pointer to the context buffer must be 8-bytes aligned.
-    _cctx_raw = allocate_aligned_buffer<char>(cctx_size, 8);
-    _cctx = ZSTD_initStaticCCtx(_cctx_raw.get(), cctx_size);
-    if (!_cctx) {
-        throw std::runtime_error("Unable to initialize ZSTD compression context");
+
+    if (cctx_size > _shared_cctx_size) [[unlikely]] {
+        // According to the ZSTD documentation, pointer to the context buffer must be 8-bytes aligned.
+        auto new_buf = allocate_aligned_buffer<char>(cctx_size, 8);
+        if (!ZSTD_initStaticCCtx(new_buf.get(), cctx_size)) {
+            // This should never happen.
+            throw std::runtime_error("Unable to initialize ZSTD compression context");
+        }
+        if (cctx_size <= max_shared_cctx_size) {
+            _shared_cctx_size = cctx_size;
+            _shared_cctx = std::move(new_buf);
+        } else {
+            _private_cctx = std::move(new_buf);
+        }
     }
 
-    auto dctx_size = ZSTD_estimateDCtxSize();
-    _dctx_raw = allocate_aligned_buffer<char>(dctx_size, 8);
-    _dctx = ZSTD_initStaticDCtx(_dctx_raw.get(), dctx_size);
-    if (!_cctx) {
-        throw std::runtime_error("Unable to initialize ZSTD decompression context");
+    if (!_shared_dctx) {
+        auto dctx_size = ZSTD_estimateDCtxSize();
+        _shared_dctx = allocate_aligned_buffer<char>(dctx_size, 8);
+        if (!ZSTD_initStaticCCtx(_shared_dctx.get(), dctx_size)) {
```

Comment:
I think there's a typo here, unless I'm minunderstanding something. Shouldn't this be calling **DC** not *CC* here?

Reply:
Yes, it's a typo. Apparently the init functions mostly just zero some stuff, so it somehow worked. You just saved me from a possibly very sneaky bug, thanks.

Reply:
By the way, there was also a similar typo before the patch in this place:
```
    _dctx = ZSTD_initStaticDCtx(_dctx_raw.get(), dctx_size);
    if (!_cctx) {
```
instead of:
```
    _dctx = ZSTD_initStaticDCtx(_dctx_raw.get(), dctx_size);
    if (!_dctx) {
```
though this `if` should never be taken either way.

I did my best not to repeat such a typo, but well, here we are...